File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-5.kt
class Solution {
    fun characterReplacement(s: String, k: Int): Int {
        val freq = IntArray(26)
        var left = 0
        var maxFreq = 0
        var longest = 0

        for (right in s.indices) {
            freq[s[right] - 'A']++
            maxFreq = maxOf(maxFreq, freq[s[right] - 'A'])

            // window size - most frequent count = replacements needed
            val windowSize = right - left + 1
            if (windowSize - maxFreq > k) {
                freq[s[left] - 'A']--
                left++
            }

            longest = maxOf(longest, right - left + 1)
        }

        return longest
    }
}

File: Data Structures & Algorithms/permutation-string/submission-5.swift
class Solution {
    func checkInclusion(_ s1: String, _ s2: String) -> Bool {
        
        guard s2.count >= s1.count else { return false }

        let windowSize = s1.count
        let s1Sorted = s1.sorted()
        let s2Array = Array(s2)

        for i in 0...(s2.count - windowSize) {
            
            let window = s2Array[i..<(i + windowSize)].sorted()
            if window == s1Sorted { return true }

        }

        return false 
    }
}
